Objectize ActionBar

gsingh93 opened the pull request.

gsingh93 commented on src/widgets/action_bar.rs:
```diff
@@ -4,52 +4,71 @@
+impl<O: Upcast<ActionBar>> ActionBarExt for O {
+    fn get_center_widget<T: Upcast<Widget>>(&self) -> Option<T> {
+        unsafe {
+            let tmp = from_glib_none(ffi::gtk_action_bar_get_center_widget(
```
I thought `from_glib_none` could return an `Option<T>` itself, but it wasn't working.

gkoz replied:
I make such functions simply return `Widget` and let the caller downcast it to whatever they expect it to be. Only the `Builder` tries to accommodate any buildable types generically because it is versatile, while these aren't expected to return 50 types of widgets.

gkoz replied:
Well okay, maybe these can return 50 types actually. The caller still is perfectly capable do do the downcasting themselves.

gkoz commented:
Thanks for contributing. It's a bit frustrating to do all this single-handedly.

gkoz commented on src/lib.rs:
```diff
@@ -210,6 +210,9 @@ pub use widgets::orientable::Orientable;
 pub use widgets::widget::Widget;
 pub use window::Window;
+#[cfg(feature = "gtk_3_12")]
+pub use widgets::action_bar::ActionBar;
```
I try to keep this list sorted alphabetically.

gkoz commented on src/widgets/mod.rs:
```diff
@@ -13,6 +13,8 @@ pub mod header_bar;
 pub mod scrollable;
 pub mod widget;
+#[cfg(feature = "gtk_3_12")]
+pub mod action_bar;
```
The same about alphabetical order.

gkoz commented:
Here you probably don't have to go to the length of making a trait because there aren't any subclasses of the `ActionBar`. For now I don't make traits for such widgets. This may change if we allow subclassing the widgets.

But if you do make a trait, you need to also add it to the `prelude` module.

gsingh93 commented:
I addressed most of the comments. I like leaving things open to extensibility and the possibility of subclassing later on, so I left the trait, let me know if you want me to remove it.

gkoz commented:
Remove it please. I'd like to keep this consistent and creating over a 100 traits without a need doesn't seem right. If subclassing support ever materializes it may not actually require them.

gkoz commented on src/prelude.rs:
```diff
@@ -17,6 +17,8 @@ pub use mvc::cell_renderer::{
 };
 pub use mvc::tree_model::TreeModelExt;
+#[cfg(feature = "gtk_3_12")]
+pub use widgets::action_bar::ActionBarExt;
```
This must be a leftover.

gsingh93 replied:
Yea, and I did `cargo build` instead of `cargo build --features gtk_3_12`, so I didn't catch this.

gkoz replied:
Heh, I may need to reenable travis for this branch :)

gkoz commented on src/widgets/action_bar.rs:
```diff
@@ -4,52 +4,55 @@
-            Some(::FFIWidget::wrap_widget(tmp_pointer))
+    pub fn get_center_widget(&self) -> Option<Widget> {
+        unsafe {
+            from_glib_none(ffi::gtk_action_bar_get_center_widget(self.upcast().to_glib_none().0));
```
Since it's an inherent method now, `upcast()` of `self` is redundant. Same below.

gsingh93 replied:
Fixed.

gkoz commented on src/widgets/action_bar.rs:
```diff
@@ -4,52 +4,53 @@
 
 //! Hide and show with animation
```
The last little thing. Since we're reexporting everything and I'm likely going to unpublish the modules, this description will be more useful at the `ActionBar` typedef.

gsingh93 replied:
Ok, done.

gkoz commented:
Great. Thanks!

I'm converting Fixed, FlowBox[Child], Grid, IconView, Layout, NoteBook, Socket, Paned, Stack[Switcher] so don't take those.
